src/Core/src/Platform/Android/MauiMaterialButton.cs

```diff
@@ -1,35 +1,169 @@
-﻿using Android.Content;
+﻿using System;
+using Android.Content;
+using Android.Graphics.Drawables;
+using Android.Views;
 using AndroidX.Core.Widget;
 using Google.Android.Material.Button;
 
 namespace Microsoft.Maui.Platform
 {
 	public class MauiMaterialButton : MaterialButton
 	{
-		// Currently Material doesn't have any bottom gravity options
-		// so we just move the layout to the bottom using
-		// SetCompoundDrawablesRelative during Layout
-		internal const int IconGravityBottom = 9999;
-		public MauiMaterialButton(Context context) : base(context)
+		// The default MaterialButton currently does not have a concept of bottom
+		// gravity which we need for .NET MAUI.
+		// In order to get this feature, we have added a custom gravity option
+		// that serves as a flag to indicate that the icon should be placed at
+		// the bottom.
+		// The real gravity value is IconGravityTop in order to perform all the
+		// normal layout calculations. We then set ForceBottomIconGravity for our
+		// custom layout pass where we simply swap the icon from the top to the
+		// bottom using SetCompoundDrawablesRelative.
+		internal const int IconGravityBottom = 0x1000;
+
+		public MauiMaterialButton(Context context)
+			: base(context)
 		{
 		}
 
-		protected override void OnLayout(bool changed, int left, int top, int right, int bottom)
+		public override int IconGravity
 		{
-			// These are hacks that seem to force the button to measure correctly
-			// when using top or bottom positioning. 
-			if (IconGravity == IconGravityBottom)
+			get => base.IconGravity;
+			set
 			{
-				var drawable = TextViewCompat.GetCompoundDrawablesRelative(this)[3];
-				drawable?.SetBounds(0, 0, drawable.IntrinsicWidth, drawable.IntrinsicHeight);
+				// Intercept the gravity value and set the flag if it's bottom.
+				ForceBottomIconGravity = value == IconGravityBottom;
+				base.IconGravity = ForceBottomIconGravity ? IconGravityTop : value;
 			}
-			else if (IconGravity == MaterialButton.IconGravityTop)
+		}
+
+		internal bool ForceBottomIconGravity { get; private set; }
+
+		protected override void OnMeasure(int widthMeasureSpec, int heightMeasureSpec)
+		{
+			if (Icon is MauiResizableDrawable currentIcon)
 			{
-				var drawable = TextViewCompat.GetCompoundDrawablesRelative(this)[1];
-				drawable?.SetBounds(0, 0, drawable.IntrinsicWidth, drawable.IntrinsicHeight);
+				// if there is BOTH an icon AND text, but the text layout has NOT been measured yet,
+				// we need to measure JUST the text first to get the remaining space for the icon
+				if (Layout is null && TextFormatted?.Length() > 0)
+				{
+					// remove the icon and measure JUST the text
+					Icon = null;
+					base.OnMeasure(widthMeasureSpec, heightMeasureSpec);
+
+					// restore the icon
+					Icon = currentIcon;
+				}
```

Comment:
Due to the way TextView measures, it appears to prioritize the icon and will clip/squash the text. However, we want the text to be the priority. In order for this to happen, we have to measure the initial pass without an icon and this sets us up to re-measure with a smaller/fitting icon.